pkg/controller/common/configref.go

```diff
@@ -0,0 +1,75 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package common
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	commonv1 "github.com/elastic/cloud-on-k8s/pkg/apis/common/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/driver"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/events"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/settings"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/watches"
+)
+
+// ConfigRefWatchName returns the name of the watch registered on the secret referenced in `configRef`.
+func ConfigRefWatchName(resource types.NamespacedName) string {
+	return fmt.Sprintf("%s-%s-configref", resource.Namespace, resource.Name)
+}
+
+// ParseConfigRef retrieves the content of a secret referenced in `configRef`, sets up dynamic watches for that secret,
+// and parses the secret content into a CanonicalConfig.
+func ParseConfigRef(
+	driver driver.Interface,
+	resource runtime.Object, // eg. Beat, EnterpriseSearch
+	configRef *commonv1.ConfigSource,
+	secretKey string, // retrieve config data from that entry in the secret
+) (*settings.CanonicalConfig, error) {
+	resourceMeta, err := meta.Accessor(resource)
+	if err != nil {
+		return nil, err
+	}
+	namespace := resourceMeta.GetNamespace()
+	resourceNsn := types.NamespacedName{Namespace: namespace, Name: resourceMeta.GetName()}
+
+	// ensure watches match the referenced secret
+	var secretNames []string
+	if configRef != nil && configRef.SecretName != "" {
+		secretNames = append(secretNames, configRef.SecretName)
+	}
+	if err := watches.WatchUserProvidedSecrets(resourceNsn, driver.DynamicWatches(), ConfigRefWatchName(resourceNsn), secretNames); err != nil {
+		return nil, err
+	}
+
+	if len(secretNames) == 0 {
+		// no secret referenced, nothing to do
+		return settings.NewCanonicalConfig(), nil
```

Comment:
Shouldn't this return `nil, nil`?

Reply:
In the enterprise search reconciliation we're doing:

```
// merge with user settings last so they take precedence
err = cfg.MergeWith(reusedCfg, tlsCfg, associationCfg, userProvidedCfg, userProvidedSecretCfg)
```

where any of those config can just be empty.  I think it feels clean that if there's no config to parse, and no error, you still get a valid empty result, for which you don't have to check nil pointers.

Reply:
I think it's nice to be able to rely on the fact that `nil` means nothing was there (vs something was there but was empty).
If we don't want to return nils maybe we should return struct instead of a pointer?
If `.MergeWith` is the only concern then I'd go with `nil, nil` since it handles nils just fine:
```
func (c *CanonicalConfig) MergeWith(cfgs ...*CanonicalConfig) error {
	for _, c2 := range cfgs {
		if c2 == nil {
			continue
		}
...
```

Reply:
👍 I'm fine with that, I'll make the change to `nil`.